### PR TITLE
[Fix][Relax][ONNX] Cast BatchNorm params to input dtype

### LINKED ISSUE #19977: [Bug] Relax ONNX frontend fails to import mixed-dtype BatchNormalization
### Expected behavior

The ONNX model should be imported successfully by the Relax ONNX frontend.

### Actual behavior

`from_onnx` fails while converting the `BatchNormalization` node because
`relax.nn.batch_norm` requires all inputs to have the same dtype.

```text
tvm.error.InternalError:
relax.nn.batch_norm requires all the input tensors to have the same dtype.
However, the gamma has dtype float32, which is different from the input
data's dtype float16.
```
### Environment
```text
TVM: 0.25.dev0, built from source
Python: 3.10
OS: Ubuntu 22.04
Target: llvm (CPU)
ONNX frontend: tvm.relax.frontend.onnx.from_onnx
```
### Steps to reproduce

```python
import numpy as np
import onnx
from onnx import TensorProto, helper, numpy_helper
from tvm.relax.frontend.onnx import from_onnx
data = helper.make_tensor_value_info(
    "data",
    TensorProto.FLOAT16,
    [1, 3, 2, 2],
)
output = helper.make_tensor_value_info(
    "output",
    TensorProto.FLOAT16,
    [1, 3, 2, 2],
)
params = [
    numpy_helper.from_array(
        np.array([1.0, 1.5, 2.0], dtype=np.float32),
        name="gamma",
    ),
    numpy_helper.from_array(
        np.array([0.0, 0.1, -0.1], dtype=np.float32),
        name="beta",
    ),
    numpy_helper.from_array(
        np.array([0.2, -0.3, 0.4], dtype=np.float32),
        name="mean",
    ),
    numpy_helper.from_array(
        np.array([1.0, 1.5, 2.0], dtype=np.float32),
        name="var",
    ),
]
node = helper.make_node(
    "BatchNormalization",
    inputs=["data", "gamma", "beta", "mean", "var"],
    outputs=["output"],
    epsilon=1e-5,
    momentum=0.9,
    training_mode=0,
)
graph = helper.make_graph(
    [node],
    "mixed_dtype_batchnorm",
    [data],
    [output],
    initializer=params,
)
model = helper.make_model(
    graph,
    opset_imports=[helper.make_opsetid("", 15)],
)
onnx.checker.check_model(model, full_check=True)
mod = from_onnx(
    model,
    keep_params_in_input=False,
)
```
### Analysis
The Relax ONNX frontend directly converts the ONNX `BatchNormalization` node into `relax.nn.batch_norm`.
The ONNX model contains mixed floating-point dtypes:
```text
data:float16, gamma:float32, beta:float32, moving_mean: float32, moving_var: float32
```
The Relax `batch_norm` operator currently requires all five inputs to have the same dtype. Therefore, the generated Relax expression fails during `BlockBuilder.normalize()` before the ONNX model can be imported.

### PR DESCRIPTION
**Pull request by viiccwen**
Fixes #19977.

ONNX `BatchNormalization` allows the input/output tensor dtype, scale/bias dtype, and mean/variance dtype to be separate floating-point type parameters.

For example, a valid ONNX model may use `float16` data with `float32` gamma, beta, mean, and variance tensors.

The Relax `batch_norm` operator currently requires all five input tensors to have the same dtype.  The ONNX frontend previously forwarded the ONNX inputs directly to `relax.nn.batch_norm`, causing import to fail during normalization
for mixed-dtype ONNX models.

This patch casts the ONNX BatchNormalization parameter tensors (`scale`, `bias`, `mean`, and `var`) to the data tensor dtype before calling Relax `batch_norm`.

This preserves the ONNX output dtype, which follows the input data dtype, while keeping the fix localized to the frontend compatibility layer.

The regression test builds a minimal ONNX BatchNormalization graph with `float16` data and `float32` parameters, imports it through the Relax ONNX frontend, and checks that the generated Relax `batch_norm` call receives same-dtype inputs.

Verification:

- `python -m pytest tests/python/relax/test_frontend_onnx.py::test_batch_norm_mixed_dtype_params tests/python/relax/test_frontend_onnx.py::test_batch_norm_defaults_to_inference_mode -q`

### CODE REVIEW

**Review by tlopex — CHANGES_REQUESTED** [review]
Thanks for fixing! But could you avoid casting the float32 parameters down to the data dtype, as this can lose precision and change the result.

Since Relax batch_norm requires matching dtypes, please promote all five inputs to a common computation dtype (float32 for this case), run batch_norm, and cast Y back to the original data dtype.

The test can verify that the batch_norm inputs are float32 while the final output remains float16 I think

**Comment by viiccwen** [comment]
> Thanks for fixing! But could you avoid casting the float32 parameters down to the data dtype, as this can lose precision and change the result.
> 
> Since Relax batch_norm requires matching dtypes, please promote all five inputs to a common computation dtype (float32 for this case), run batch_norm, and cast Y back to the original data dtype.
> 
> The test can verify that the batch_norm inputs are float32 while the final output remains float16 I think

Thx for pointing out! I'll fix it later. : D

**Comment by viiccwen** [comment]
Hello @tlopex, 

Updated the implementation. (Follow by [ONNX BatchNormalization-15 schema](https://onnx.ai/onnx/operators/onnx__BatchNormalization.html))

The strategy is:
- Preserve ONNX's dtype groups: X/Y use T, scale/bias use T1, and mean/variance outputs use T2.
- For the reported mixed-precision case (float16 X with float16/float32 parameters and statistics), run Relax batch_norm in float32.
- Cast Y back to T and the running mean/variance back to T2.
- In training mode, float16 inputs also use float32 computation, following the [ONNX requirement](https://github.com/onnx/onnx/blob/1f5f9ef9f9714c12da0adf79e2cbd49a35dab1db/onnx/defs/nn/defs.cc#L1644-L1660) to compute training statistics in float to avoid overflow.
- Other mixed-dtype combinations are rejected explicitly instead of applying an unspecified general promotion policy. (can be discussed how to handle. : D)

**Review by tlopex — APPROVED** [review]
LGTM
